### PR TITLE
io: make copy cooperative

### PR DESCRIPTION
**Pull request by hi-rustin**
## Motivation

close https://github.com/tokio-rs/tokio/issues/6250

## Solution

Add poll_proceed_and_make_progress to check if it exceeds the budget.

### CODE REVIEW

File: tokio/tests/io_copy_bidirectional.rs [hunk 1]
```diff
@@ -138,3 +138,28 @@ async fn immediate_exit_on_read_error() {
 
     assert!(copy_bidirectional(&mut a, &mut b).await.is_err());
 }
+
+#[tokio::test]
+async fn copy_bidirectional_is_cooperative() {
```
Comment (hi-rustin) [inline tokio/tests/io_copy_bidirectional.rs:143]: It seems the copy_bidirectional is already cooperative before I make the CopyBuffer cooperative. I'm not sure why. Could I be using the wrong case to test it?

**Review by hi-rustin — COMMENTED** [review]
🔢 Self-check

File: tokio/src/io/util/copy.rs [hunk 2]
```diff
@@ -82,6 +83,8 @@ impl CopyBuffer {
         R: AsyncRead + ?Sized,
         W: AsyncWrite + ?Sized,
     {
+        ready!(crate::trace::trace_leaf(cx));
+        ready!(poll_proceed_and_make_progress(cx));
```
Comment (Darksonn) [inline tokio/src/io/util/copy.rs:87]: You should only make progress if it actually copied data.

See the mpsc channel for an example:

https://github.com/tokio-rs/tokio/blob/7c606ab44aa9f0c33dfcf5bc8678411dddeaa9e5/tokio/src/sync/mpsc/chan.rs#L250-L251

https://github.com/tokio-rs/tokio/blob/7c606ab44aa9f0c33dfcf5bc8678411dddeaa9e5/tokio/src/sync/mpsc/chan.rs#L259-L263

Comment (hi-rustin) [reply tokio/src/io/util/copy.rs]: Thank you for your review.
How about now? After filling, writing, and flushing data, I called `make_progress`.

File: tokio/tests/io_copy_bidirectional.rs (hunk 1 above)
Comment (Darksonn) [reply tokio/tests/io_copy_bidirectional.rs:143]: You should probably change this test to use something different than `tokio_test::io::Builder` to fix that.

File: tokio/src/io/util/copy.rs (hunk 2 above)
Comment (Darksonn) [reply tokio/src/io/util/copy.rs]: Like in the examples I linked, you need to call `crate::runtime::coop::poll_proceed` at the start of the function, and then call `coop.made_progress()` in the places where you are currently calling `poll_proceed_and_make_progress`.

It's not enough to just call it when it exits. The `poll_proceed_and_make_progress` function can only be used for functions that never return `Poll::Pending`.

Comment (hi-rustin) [reply tokio/src/io/util/copy.rs]: > Like in the examples I linked, you need to call `crate::runtime::coop::poll_proceed` at the start of the function, and then call `coop.made_progress()`

But the crate::runtime::coop::poll_proceed only exists if we enable the coop feature. What should we do if we don't have the `coop` feature?

File: tokio/tests/io_copy_bidirectional.rs (hunk 1 above)
Comment (hi-rustin) [reply tokio/tests/io_copy_bidirectional.rs:143]: I tried to use `symmetric`, but it also didn't work.

File: tokio/src/io/util/copy.rs (hunk 2 above)
Comment (Darksonn) [reply tokio/src/io/util/copy.rs]: Then do nothing.

Comment (hi-rustin) [reply tokio/src/io/util/copy.rs]: Updated. I just used `#[cfg(any(
                    feature = "fs",
                    feature = "io-std",
                    feature = "net",
                    feature = "process",
                    feature = "rt",
                    feature = "signal",
                    feature = "sync",
                    feature = "time",
                    ))]` to control it. I am not sure if we have a better way to do it. Thanks for your help. Sorry for those stupid questions. 😢

Comment (Darksonn) [reply tokio/src/io/util/copy.rs]: I believe we have [a macro](https://github.com/tokio-rs/tokio/blob/2f730d4d5a4810088428f6e6397bda42d376deec/tokio/src/macros/cfg.rs#L487-L503) for doing that. You should be able to write
```rs
cfg_coop! {
    let coop = ready!(crate::runtime::coop::poll_proceed(cx));
}
```
And similar for the calls to `make_progress`. The macro does not actually introduce a scope, so I believe this should work.

Your questions are not stupid :). You're asking questions about internal Tokio APIs that don't exist elsewhere. It's no surprise that you are not familiar with them. If we did not have the `cfg_coop!` macro, then what you did would be the only way to do it.

Comment (hi-rustin) [reply tokio/src/io/util/copy.rs]: Thanks for your reply.

Actually, I tried it. But it seemed this macro only supports starting with an item keyword. 

Do you think I should extend it to support it? (I will try, but I am unsure if it is doable).

**Review by Darksonn — APPROVED** [review]
Thank you. If the `cfg_coop!` macro can't do this today, then this is good enough.

**Comment by hi-rustin** [comment]
Thanks for your review!  💚 💙 💜 💛 ❤️

**Pull request merged**